File: npm-core/openplugincore/src/openpluginMemo.ts
import { OpenPlugin } from './openPlugin';

export class OpenPluginMemo {
  plugins: Map<string, OpenPlugin>;
  pluginsDirectory: Map<string, string> | undefined;

  constructor() {
    this.plugins = new Map();
    this.pluginsDirectory;
  }

  async init() {
    this.pluginsDirectory = new Map();
    const plugins_url =
      'https://raw.githubusercontent.com/CakeCrusher/openplugin/main/migrations/plugin_store/openplugins.json';
    const response = await fetch(plugins_url);
    if (!response.ok) {
      throw new Error(
        `Unable to fetch plugins from github url '${plugins_url}'`
      );
    }
    // add all json key and values to pluginsDirectory
    const plugins = await response.json();
    for (const [key, value] of Object.entries(plugins)) {
      this.pluginsDirectory.set(key, value as string);
    }
    console.log('MEMO READY');
  }

  async initPlugin(pluginName: string) {
    if (!this.pluginsDirectory) {
      throw new Error('Plugin directory not initialized');
    }
    const pluginUrl = this.pluginsDirectory.get(pluginName);
    if (pluginUrl === undefined) {
      throw new Error('Plugin not found');
    }
    const plugin = new OpenPlugin(
      pluginName,
      pluginUrl,
      process.env.OPENAI_API_KEY
    );
    await plugin.init();
    this.addPlugin(plugin);
    return plugin;
  }

  // method to add an OpenPlugin to the plugins map
  addPlugin(plugin: OpenPlugin) {
    if (!plugin.pluginName) throw new Error('Plugin name not found');
    this.plugins.set(plugin.pluginName, plugin);
  }

  // method to remove an OpenPlugin from the plugins map
  removePlugin(pluginName: string) {
    this.plugins.delete(pluginName);
  }

  // method to get an OpenPlugin by name
  getPlugin(pluginName: string) {
    return this.plugins.get(pluginName);
  }
}

File: pypi-core/openplugincore/openplugin_memo.py
import requests
from openplugincore import OpenPlugin
import os

class OpenPluginMemo:
    def __init__(self):
        self.plugins = {}
        self.plugins_directory = None

    def init(self):
        self.plugins_directory = {}
        plugins_url = 'https://raw.githubusercontent.com/CakeCrusher/openplugin/main/migrations/plugin_store/openplugins.json'
        response = requests.get(plugins_url)
        if not response.ok:
            raise Exception(f"Unable to fetch plugins from github url '{plugins_url}'")
        # add all json key and values to pluginsDirectory
        plugins = response.json()
        for key, value in plugins.items():
            self.plugins_directory[key] = value
        print('MEMO READY')

    def init_plugin(self, plugin_name):
        if not self.plugins_directory:
            raise Exception('Plugin directory not initialized')
        plugin_url = self.plugins_directory.get(plugin_name)
        if plugin_url is None:
            raise Exception('Plugin not found')
        plugin = OpenPlugin(plugin_name=plugin_name, openai_api_key=os.getenv('OPENAI_API_KEY'))
        plugin.init()
        self.add_plugin(plugin)
        return plugin

    # method to add an OpenPlugin to the plugins map
    def add_plugin(self, plugin):
        if not plugin.name:
            raise Exception('Plugin name not found')
        self.plugins[plugin.name] = plugin

    # method to remove an OpenPlugin from the plugins map
    def remove_plugin(self, plugin_name):
        self.plugins.pop(plugin_name, None)

    # method to get an OpenPlugin by name
    def get_plugin(self, plugin_name):
        return self.plugins.get(plugin_name)
